src/Zforce.cpp

```diff
@@ -123,6 +124,110 @@ int Zforce::Write(uint8_t* payload)
+ *                Sadly, some Arduino i2c drivers do not signal errors.
+ */
+bool Zforce::SendRawMessage(uint8_t* payload, uint8_t payloadLength)
+{
+  if ((payload == nullptr) || (payloadLength == 0))
+  {
+    return false;
+  }
+
+  buffer[0] = 0xEE;
+  buffer[1] = payloadLength;
+  memcpy(&buffer[2], payload, payloadLength);
+
+  return Write(buffer) == 0;
+}
+
+/*
+ * Receive a raw ASN.1 message. The only validation done is decoding the initial ASN.1 payload length
+ * to see if more data should follow.
+ *
+ * receivedLength   A pointer to where the size of the returned data should be placed.
+ * remainingLength  A pointer to where the size of the remaining data should be placed, if any.
+ *
+ * Return value     If an error occured (see comment above for the return value of SendRawMessage),
+ *                  the pointer will be nullptr.
+ *                  A pointer to the ASN.1 payload of the received data is returned.
+ *                  Note, that any subsequent read or write-operation, even notifications, touches, etc will
+ *                  will overwrite the buffer, so make sure to copy any data you want to save.
+ */
+uint8_t* Zforce::ReceiveRawMessage(uint8_t* receivedLength, uint16_t *remainingLength)
+{
+  if ((GetDataReady() == HIGH) && !Read(buffer))
+  {
+    uint8_t i2cPayloadLength = buffer[1];
+    // Check if this is a second or first call.
+    if (this->remainingRawLength == 0)
+    {
+      // This is the first invocation.
+      if (buffer[2] != 0xEF)
+      {
+        // This was not a response.
+        *receivedLength = 0;
+        *remainingLength = 0;
+        return nullptr;
+      }
+      // Check if it's a short, 2, or 3 byte length encoding.
+      uint8_t firstLengthByte = buffer[3];
+      uint16_t asn1AfterHeaderLength;
+      uint8_t asn1HeaderLength = 2; // EE/EF/F0 + First byte of length.
+      if (firstLengthByte < 0x80)
+      {
+        // Short form. Lower 7 bits are the length, but since it's 0, we don't need to & 0x7F to get it.
+        asn1AfterHeaderLength = firstLengthByte;
+      }
+      else
+      {
+        // Long form. First byte's top bit is set. The lower 7 bits contain the number of length bytes.
+        // The following 1 or 2 bytes contain the actual length, in Big Endian / Intel Byte Order encoding.
+        uint8_t numberOfLengthBytes = (firstLengthByte & 0x7F);
+        asn1HeaderLength += numberOfLengthBytes;
+        asn1AfterHeaderLength = buffer[4];
+        if (numberOfLengthBytes == 2)
+        {
+          asn1AfterHeaderLength <<= 8;
+          asn1AfterHeaderLength += buffer[5];
+        }
+        else if (numberOfLengthBytes != 1)
+        {
+          // The data is most likely corrupted. Valid numbers are 1 and 2.
+          *receivedLength = 0;
+          *remainingLength = 0;
+          return nullptr;
+        }
+      }
+      uint16_t fullAsn1MessageLength = asn1AfterHeaderLength + asn1HeaderLength
```

Comment:
It still doesn't compile. One might think the Arduino GCC compiler is picky but it is missing a semicolon.